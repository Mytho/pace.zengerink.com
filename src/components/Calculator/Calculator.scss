$border-radius: 4px;

@mixin calculator-row {
    display: flex;
    justify-content: space-between;
    padding: 13px;

    &__label {
        color: #555;
    }

    &__value-label {
        margin-left: 3px;
        color: #333;
    }
}

.calculator {
    width: 300px;
    margin: 20px auto;
    border-radius: $border-radius;
    box-shadow: 6px 8px 42px 1px rgba(0, 0, 0, .15);

    &__header {
        padding: 13px 0 13px 16px;
        background-color: #e6e6e6;
        color: #333;
        border-top-left-radius: $border-radius;
        border-top-right-radius: $border-radius;
    }

    &:last-child {
        border-bottom-left-radius: $border-radius;
        border-bottom-right-radius: $border-radius;
    }
}