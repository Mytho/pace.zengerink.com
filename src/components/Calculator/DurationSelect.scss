@import './Calculator.scss';

.duration-select {
    @include calculator-row;
}