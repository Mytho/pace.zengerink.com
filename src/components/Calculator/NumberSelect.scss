@import './Calculator.scss';

.number-select {
    &__wrapper {
        float: left;
        display: inline;
        border-bottom: 1px solid $highlight-dark;
        cursor: pointer;
    }

    &__value {
        font-weight: 600;
    }

    &__options {
        display: none;
        position: absolute;
        height: 250px;
        background-color: #fff;
        overflow-y: scroll;
        box-shadow: $box-shadow;
        border-bottom-left-radius: $border-radius;
        border-bottom-right-radius: $border-radius;
    }

    &__option {
        padding: 4px 15px 4px 4px;
        cursor: pointer;

        &--selected {
            font-weight: 600;
            background-color: $highlight-light;
        }

        &:hover {
            background-color: $hover-grey;
        }
    }

    &--active &__options {
        border-top: 1px solid $highlight-dark;
        display: block;
    }
}