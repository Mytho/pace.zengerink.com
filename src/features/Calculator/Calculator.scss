@import '../../App.scss';

@mixin calculator-row {
    display: flex;
    justify-content: space-between;
    padding: 13px 0 13px 13px;
    border-bottom: 1px solid $light-grey;
    background-color: $white;

    &__label {
        color: $medium-grey;
    }

    &__input {
        flex-basis: 180px;
    }

    &__separator,
    &__value-label {
        float: left;
        color: $dark-grey;
    }

    &__separator {
        margin: 0 1px;
    }

    &__value-label {
        margin: 0 7px 0 4px;
    }

    &:last-child {
        border-bottom: 0;
        border-bottom-left-radius: $border-radius;
        border-bottom-right-radius: $border-radius;
    }
}

.calculator {
    width: 280px;
    margin: 40px auto;
    border-radius: $border-radius;
    box-shadow: $box-shadow;

    &__header {
        display: flex;
        justify-content: space-between;
        padding: 13px;
        background-color: $light-grey;
        color: $medium-light-grey;
        border-top-left-radius: $border-radius;
        border-top-right-radius: $border-radius;
    }

    &__header div {
        margin-top: 2px;
    }

    &__header button {
        padding: 4px;
        color: $medium-light-grey;
        border: 0;
        border-radius: 3px;
        background-color: $lighter-grey;
        cursor: pointer;

        &:focus {
            outline: none;
        }

        &:active,
        &:hover {
            color: $dark-highlight;
            background-color: $light-highlight;
        }
    }
}
