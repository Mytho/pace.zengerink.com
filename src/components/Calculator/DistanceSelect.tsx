import React from 'react'
import NumberSelect from './NumberSelect'
import './DistanceSelect.scss'

export enum DistanceFragment {
  Kilometers = 'km',
  Meters = 'm',
}

interface DistanceSelectProps {
  label: string
  value: number
  onChange: (value: number) => void
}

interface Distance {
  [key: string]: number
}

export default class DurationSelect extends React.Component<DistanceSelectProps, {}> {
  handleChange(prop: string, value: number) {
    const distance = this.toDistance(this.props.value)
    distance[prop] = value
    this.props.onChange(this.toMeters(distance))
  }

  toMeters(distance: Distance): number {
    return distance[DistanceFragment.Kilometers] * 1000 + distance[DistanceFragment.Meters]
  }

  toDistance(meters: number): Distance {
    return {
      [DistanceFragment.Kilometers]: Math.floor(meters / 1000),
      [DistanceFragment.Meters]: meters % 1000,
    }
  }

  render() {
    const distance = this.toDistance(this.props.value)

    return (
      <div className="distance-select">
        <div className="distance-select__label">{this.props.label}</div>
        <div className="distance-select__input">
          <NumberSelect
            name={DistanceFragment.Kilometers}
            value={distance[DistanceFragment.Kilometers]}
            max="100"
            onChange={this.handleChange.bind(this, DistanceFragment.Kilometers)}
          />
          <span className="distance-select__separator">.</span>
          <NumberSelect
            name={DistanceFragment.Meters}
            value={distance[DistanceFragment.Meters]}
            max="999"
            pad="3"
            onChange={this.handleChange.bind(this, DistanceFragment.Meters)}
          />
          <span className="distance-select__value-label">km</span>
        </div>
      </div>
    )
  }
}
