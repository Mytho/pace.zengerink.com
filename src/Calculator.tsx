import React from 'react'
import DurationSelect from './DurationSelect'
import DistanceSelect from './DistanceSelect'

enum CalculatorValue {
    Seconds = 'seconds',
    Meters = 'meters'
}

interface CalculatorProps {}

interface CalculatorState {
    [key: string]: number
}

export default class Calculator extends React.Component<CalculatorProps, CalculatorState> {
    constructor(props: any) {
        super(props)

        this.state = {
            [CalculatorValue.Seconds]: 0,
            [CalculatorValue.Meters]: 0
        }
    }

    handleChange(prop: CalculatorValue, value: number) {
        this.setState({
            [prop]: value,
        })
    }

    render() {
        return (
            <div className="calculator">
                <div className="calculator__header">Pace Calculator</div>
                <DurationSelect value={this.state[CalculatorValue.Seconds]} onChange={this.handleChange.bind(this, CalculatorValue.Seconds)} />
                <DistanceSelect value={this.state[CalculatorValue.Meters]} onChange={this.handleChange.bind(this, CalculatorValue.Meters)} />
            </div>
        )
    }
}