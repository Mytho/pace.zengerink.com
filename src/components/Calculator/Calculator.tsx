import React from 'react'
import DurationSelect from './DurationSelect'
import DistanceSelect from './DistanceSelect'
import PaceSelect from './PaceSelect'
import './Calculator.scss'

export enum CalculatorValue {
  DurationInSeconds = 'seconds',
  DistanceInMeters = 'meters',
  PaceInSeconds = 'pace',
}

interface CalculatorState {
  seconds: number
  meters: number
  pace: number
  changed: CalculatorValue[]
}

export default class Calculator extends React.Component<{}, CalculatorState> {
  constructor(props: any) {
    super(props)

    this.state = {
      [CalculatorValue.DurationInSeconds]: 0,
      [CalculatorValue.DistanceInMeters]: 0,
      [CalculatorValue.PaceInSeconds]: 0,
      changed: [],
    }
  }

  calculateDistance(duration: number, pace: number): number {
    const durationInMinutes = duration / 60
    const paceInMinutes = pace / 60
    const paceInKilometers = durationInMinutes / paceInMinutes
    return paceInKilometers * 1000
  }

  calculateDuration(distance: number, pace: number): number {
    const distanceInKilometers = distance / 1000
    const paceInMinutes = pace / 60
    const durationInMinutes = distanceInKilometers * paceInMinutes
    return durationInMinutes * 60
  }

  calculatePace(duration: number, distance: number): number {
    const distanceInKilometers = distance / 1000
    const durationInMinutes = duration / 60
    const paceInMinutes = durationInMinutes / distanceInKilometers
    return Math.round(paceInMinutes * 60)
  }

  calculateStateProp(state: CalculatorState, prop: CalculatorValue, calc: () => number): CalculatorState {
    if (state[prop] === 0 || !state.changed.find((item) => item === prop)) {
      const value = calc()
      state[prop] = Number.isFinite(value) ? value : 0
    }
    return state
  }

  handleChange(prop: CalculatorValue, value: number) {
    this.setState((prevState) => {
      let nextState = {
        [CalculatorValue.DurationInSeconds]: prevState[CalculatorValue.DurationInSeconds],
        [CalculatorValue.DistanceInMeters]: prevState[CalculatorValue.DistanceInMeters],
        [CalculatorValue.PaceInSeconds]: prevState[CalculatorValue.PaceInSeconds],
        changed: prevState.changed,
      }
      nextState[prop] = value

      // Push current item to the `changed` list, only keep two recent items
      if (!prevState.changed.some((item) => item === prop)) {
        if (nextState.changed.push(prop) > 2) {
          nextState.changed.shift()
        }
      }

      nextState = this.calculateStateProp(nextState, CalculatorValue.DurationInSeconds, () => {
        return this.calculateDuration(
          nextState[CalculatorValue.DistanceInMeters],
          nextState[CalculatorValue.PaceInSeconds]
        )
      })

      nextState = this.calculateStateProp(nextState, CalculatorValue.DistanceInMeters, () => {
        return this.calculateDistance(
          nextState[CalculatorValue.DurationInSeconds],
          nextState[CalculatorValue.PaceInSeconds]
        )
      })

      nextState = this.calculateStateProp(nextState, CalculatorValue.PaceInSeconds, () => {
        return this.calculatePace(
          nextState[CalculatorValue.DurationInSeconds],
          nextState[CalculatorValue.DistanceInMeters]
        )
      })

      return nextState
    })
  }

  render() {
    return (
      <div className="calculator">
        <div className="calculator__header">Pace Calculator</div>
        <DurationSelect
          label="Time"
          value={this.state[CalculatorValue.DurationInSeconds]}
          onChange={this.handleChange.bind(this, CalculatorValue.DurationInSeconds)}
        />
        <DistanceSelect
          label="Distance"
          value={this.state[CalculatorValue.DistanceInMeters]}
          onChange={this.handleChange.bind(this, CalculatorValue.DistanceInMeters)}
        />
        <PaceSelect
          label="Pace"
          value={this.state[CalculatorValue.PaceInSeconds]}
          onChange={this.handleChange.bind(this, CalculatorValue.PaceInSeconds)}
        />
      </div>
    )
  }
}
