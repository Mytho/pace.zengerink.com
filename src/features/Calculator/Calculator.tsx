import React, { ReactElement, useReducer } from 'react'
import DurationSelect from './DurationSelect'
import DistanceSelect from './DistanceSelect'
import PaceSelect from './PaceSelect'
import './Calculator.scss'

type CalculatorField = 'duration' | 'distance' | 'pace'

interface CalculatorState {
  duration: number
  distance: number
  pace: number
  changedFields: CalculatorField[]
}

interface CalculatorChangePayload {
  field: CalculatorField
  value: number
}

interface CalculatorAction {
  type: string
  payload?: CalculatorChangePayload
}

const calculateDistance = (duration: number, pace: number) => {
  const durationInMinutes = duration / 60
  const paceInMinutes = pace / 60
  const paceInKilometers = durationInMinutes / paceInMinutes
  return Math.round(paceInKilometers * 1000)
}

const calculateDuration = (distance: number, pace: number) => {
  const distanceInKilometers = distance / 1000
  const paceInMinutes = pace / 60
  const durationInMinutes = distanceInKilometers * paceInMinutes
  return Math.round(durationInMinutes * 60)
}

const calculatePace = (duration: number, distance: number) => {
  const distanceInKilometers = distance / 1000
  const durationInMinutes = duration / 60
  const paceInMinutes = durationInMinutes / distanceInKilometers
  return Math.round(paceInMinutes * 60)
}

const Calculator = (): ReactElement => {
  const defaultState = {
    duration: 0,
    distance: 0,
    pace: 0,
    changedFields: [],
  } as CalculatorState

  const reducer = (prevState: CalculatorState, action: CalculatorAction): CalculatorState => {
    switch (action.type) {
      case 'calculator/clear':
        return defaultState

      case 'calculator/change':
        const field = action.payload?.field ? action.payload?.field : ''
        const value = action.payload?.value

        const getValue = (fieldName: CalculatorField) => {
          const val = field === fieldName ? value : prevState[fieldName]
          return val ? val : 0
        }

        const duration = getValue('duration')
        const distance = getValue('distance')
        const pace = getValue('pace')

        const changedFields = [
          ...prevState.changedFields.filter((f: CalculatorField) => f !== field).slice(-1),
          field as CalculatorField,
        ]

        const shouldCalculate = (fieldName: CalculatorField): boolean => {
          if (changedFields.length < 2) return false
          return !changedFields.find((changedField: CalculatorField) => changedField === fieldName)
        }

        return {
          duration: shouldCalculate('duration') ? calculateDuration(distance, pace) : duration,
          distance: shouldCalculate('distance') ? calculateDistance(duration, pace) : distance,
          pace: shouldCalculate('pace') ? calculatePace(duration, distance) : pace,
          changedFields,
        }

      default:
        throw new Error()
    }
  }

  const [state, dispatch] = useReducer(reducer, defaultState)

  return (
    <div className="calculator">
      <div className="calculator__header">
        <div>Pace Calculator</div>
        <button onClick={() => dispatch({ type: 'calculator/clear' })}>Clear</button>
      </div>
      <DurationSelect
        label="Time"
        value={state?.duration}
        onChange={(value) => dispatch({ type: 'calculator/change', payload: { field: 'duration', value } })}
      />
      <DistanceSelect
        label="Distance"
        value={state?.distance}
        onChange={(value) => dispatch({ type: 'calculator/change', payload: { field: 'distance', value } })}
      />
      <PaceSelect
        label="Pace"
        value={state?.pace}
        onChange={(value) => dispatch({ type: 'calculator/change', payload: { field: 'pace', value } })}
      />
    </div>
  )
}

export default Calculator
