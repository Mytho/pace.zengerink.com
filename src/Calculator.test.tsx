import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import Calculator, { CalculatorValue } from './Calculator'

describe('render', () => {
  it('renders correctly', () => {
    const json = shallowToJson(shallow(<Calculator />))
    expect(json).toMatchSnapshot()
  })
})

describe('handleChange', () => {
  const testValues = {
    [CalculatorValue.DurationInSeconds]: 600, // 10 min
    [CalculatorValue.DistanceInMeters]: 2000, // 2 km
    [CalculatorValue.PaceInSeconds]: 300 // 5 min/km
  }

  it('calculates pace when duration and distance have been set', () => {
    const wrapper = shallow(<Calculator />)
    const instance = wrapper.instance() as Calculator
    instance.handleChange(CalculatorValue.DurationInSeconds, testValues[CalculatorValue.DurationInSeconds])
    instance.handleChange(CalculatorValue.DistanceInMeters, testValues[CalculatorValue.DistanceInMeters])

    expect(wrapper.state()).toEqual({
      changed: [CalculatorValue.DurationInSeconds, CalculatorValue.DistanceInMeters],
      [CalculatorValue.DurationInSeconds]: testValues[CalculatorValue.DurationInSeconds],
      [CalculatorValue.DistanceInMeters]: testValues[CalculatorValue.DistanceInMeters],
      [CalculatorValue.PaceInSeconds]: testValues[CalculatorValue.PaceInSeconds],
    })
  })

  it('calculates duration when distance and pace have been set', () => {
    const wrapper = shallow(<Calculator />)
    const instance = wrapper.instance() as Calculator
    instance.handleChange(CalculatorValue.DistanceInMeters, testValues[CalculatorValue.DistanceInMeters])
    instance.handleChange(CalculatorValue.PaceInSeconds, testValues[CalculatorValue.PaceInSeconds])

    expect(wrapper.state()).toEqual({
      changed: [CalculatorValue.DistanceInMeters, CalculatorValue.PaceInSeconds],
      [CalculatorValue.DurationInSeconds]: testValues[CalculatorValue.DurationInSeconds],
      [CalculatorValue.DistanceInMeters]: testValues[CalculatorValue.DistanceInMeters],
      [CalculatorValue.PaceInSeconds]: testValues[CalculatorValue.PaceInSeconds],
    })
  })

  it('calculates distance when duration and pace have been set', () => {
    const wrapper = shallow(<Calculator />)
    const instance = wrapper.instance() as Calculator
    instance.handleChange(CalculatorValue.DurationInSeconds, testValues[CalculatorValue.DurationInSeconds])
    instance.handleChange(CalculatorValue.PaceInSeconds, testValues[CalculatorValue.PaceInSeconds])

    expect(wrapper.state()).toEqual({
      changed: [CalculatorValue.DurationInSeconds, CalculatorValue.PaceInSeconds],
      [CalculatorValue.DurationInSeconds]: testValues[CalculatorValue.DurationInSeconds],
      [CalculatorValue.DistanceInMeters]: testValues[CalculatorValue.DistanceInMeters],
      [CalculatorValue.PaceInSeconds]: testValues[CalculatorValue.PaceInSeconds],
    })
  })

  it('stores the last two changed selects', () => {
    const wrapper = shallow(<Calculator />)
    const instance = wrapper.instance() as Calculator

    instance.handleChange(CalculatorValue.DurationInSeconds, testValues[CalculatorValue.DurationInSeconds])
    expect(wrapper.state()).toMatchObject({changed: [CalculatorValue.DurationInSeconds]})

    instance.handleChange(CalculatorValue.DistanceInMeters, testValues[CalculatorValue.DistanceInMeters])
    expect(wrapper.state()).toMatchObject({changed: [CalculatorValue.DurationInSeconds, CalculatorValue.DistanceInMeters]})

    instance.handleChange(CalculatorValue.PaceInSeconds, testValues[CalculatorValue.PaceInSeconds])
    expect(wrapper.state()).toMatchObject({changed: [CalculatorValue.DistanceInMeters, CalculatorValue.PaceInSeconds]})
  })

  it('calculates the value of the last unchanged select', () => {
    const wrapper = shallow(<Calculator />)
    const instance = wrapper.instance() as Calculator

    instance.handleChange(CalculatorValue.DurationInSeconds, testValues[CalculatorValue.DurationInSeconds])
    instance.handleChange(CalculatorValue.DistanceInMeters, testValues[CalculatorValue.DistanceInMeters])
    instance.handleChange(CalculatorValue.PaceInSeconds, testValues[CalculatorValue.PaceInSeconds])

    expect(wrapper.state()).toMatchObject({
      [CalculatorValue.DurationInSeconds]: testValues[CalculatorValue.DurationInSeconds],
      [CalculatorValue.DistanceInMeters]: testValues[CalculatorValue.DistanceInMeters],
      [CalculatorValue.PaceInSeconds]: testValues[CalculatorValue.PaceInSeconds],
    })

    const values = {
      [CalculatorValue.DurationInSeconds]: 300, // 5 min
      [CalculatorValue.DistanceInMeters]: 5000, // 5 km
      [CalculatorValue.PaceInSeconds]: 60 // 1 min/km
    }

    instance.handleChange(CalculatorValue.DurationInSeconds, values[CalculatorValue.DurationInSeconds])
    instance.handleChange(CalculatorValue.PaceInSeconds, values[CalculatorValue.PaceInSeconds])

    expect(wrapper.state()).toMatchObject({
      [CalculatorValue.DurationInSeconds]: values[CalculatorValue.DurationInSeconds],
      [CalculatorValue.DistanceInMeters]: values[CalculatorValue.DistanceInMeters],
      [CalculatorValue.PaceInSeconds]: values[CalculatorValue.PaceInSeconds]
    })
  })
})