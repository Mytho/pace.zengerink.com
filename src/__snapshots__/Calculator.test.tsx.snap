// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders correctly 1`] = `
<div
  className="calculator"
>
  <div
    className="calculator__header"
  >
    Pace Calculator
  </div>
  <DurationSelect
    onChange={[Function]}
    value={0}
  />
  <DurationSelect
    onChange={[Function]}
    value={0}
  />
</div>
`;
